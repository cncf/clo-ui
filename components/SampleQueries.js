module.exports = require('../dist/cjs/components/SampleQueries.js');
