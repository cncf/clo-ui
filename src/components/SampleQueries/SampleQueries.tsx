import { isUndefined, sampleSize } from 'lodash';
import React, { Fragment } from 'react';

export interface ISampleQueriesProps {
  queries: SampleQuery[];
  className?: string;
  lineBreakIn?: number;
  prepareQueryString: (items: any) => string;
  maxQueriesNumber: number;
}

export interface SampleQuery {
  name: string;
  filters: any;
}

export const SampleQueries = React.memo(function SampleQueries(props: ISampleQueriesProps) {
  const queries =
    props.queries.length > props.maxQueriesNumber ? sampleSize(props.queries, props.maxQueriesNumber) : props.queries;
  return (
    <>
      {queries.map((query: SampleQuery, index: number) => (
        <Fragment key={`sampleQuery_${index}`}>
          <a
            className={`badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none ${props.className}`}
            href={`/search${props.prepareQueryString(query.filters)}`}
            target="_self"
            rel="noopener noreferrer"
            aria-label={`Filter by ${query.name}`}
          >
            {query.name}
          </a>
          {!isUndefined(props.lineBreakIn) && index === props.lineBreakIn - 1 && (
            <div className="d-block w-100" data-testid="sampleQueryBreakLine" />
          )}
        </Fragment>
      ))}
    </>
  );
});
