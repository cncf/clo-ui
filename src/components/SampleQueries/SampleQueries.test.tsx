import { render, screen } from '@testing-library/react';

import { SampleQueries } from './SampleQueries';

const mockQueries = [
  {
    name: 'Only graduated projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['graduated'] },
    },
  },
  {
    name: 'Only incubating projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['incubating'] },
    },
  },
  {
    name: 'Only sandbox projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['sandbox'] },
    },
  },
];

const defaultProps = {
  queries: mockQueries,
  prepareQueryString: () => '?maturity=graduated&page=1',
  maxQueriesNumber: 5,
};

jest.mock('lodash', () => ({
  ...(jest.requireActual('lodash') as unknown as object),
  sampleSize: () => {
    return mockQueries;
  },
}));

describe('SampleQueries', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<SampleQueries {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(<SampleQueries {...defaultProps} />);

      expect(screen.getAllByRole('link', { name: /Filter by/ })).toHaveLength(mockQueries.length);

      for (let i = 0; i < mockQueries.length; i++) {
        expect(screen.getByText(mockQueries[i].name)).toBeInTheDocument();
      }
    });

    it('renders proper classes', () => {
      render(<SampleQueries {...defaultProps} className="custom" />);

      const links = screen.getAllByRole('link', { name: /Filter by/ });
      expect(links[0]).toHaveClass('badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none custom');
    });

    it('checks first sample query', async () => {
      render(<SampleQueries {...defaultProps} />);

      const links = screen.getAllByRole('link', { name: /Filter by/ });
      expect(links[0]).toHaveProperty('href', 'http://localhost/search?maturity=graduated&page=1');
    });
  });
});
