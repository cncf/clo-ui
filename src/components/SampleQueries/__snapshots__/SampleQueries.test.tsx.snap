// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleQueries creates snapshot 1`] = `
<DocumentFragment>
  <a
    aria-label="Filter by Only graduated projects"
    class="badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none undefined"
    href="/search?maturity=graduated&page=1"
    rel="noopener noreferrer"
    target="_self"
  >
    Only graduated projects
  </a>
  <a
    aria-label="Filter by Only incubating projects"
    class="badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none undefined"
    href="/search?maturity=graduated&page=1"
    rel="noopener noreferrer"
    target="_self"
  >
    Only incubating projects
  </a>
  <a
    aria-label="Filter by Only sandbox projects"
    class="badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none undefined"
    href="/search?maturity=graduated&page=1"
    rel="noopener noreferrer"
    target="_self"
  >
    Only sandbox projects
  </a>
</DocumentFragment>
`;
